airflow/models.py

```diff
@@ -628,14 +642,29 @@ def are_dependencies_met(self, main_session=None):
         # Checking that all upstream dependencies have succeeded
         if task._upstream_list:
             upstream_task_ids = [t.task_id for t in task._upstream_list]
-            ti = session.query(func.count(TI.task_id)).filter(
-                TI.dag_id == self.dag_id,
-                TI.task_id.in_(upstream_task_ids),
-                TI.execution_date == self.execution_date,
-                TI.state == State.SUCCESS,
+            qry = (
+                session
+                .query(
+                    func.sum(
+                        case([(TI.state==State.SUCCESS, 1)], else_=0)),
+                    func.count(TI.task_id),
+                )
+                .filter(
+                    TI.dag_id == self.dag_id,
+                    TI.task_id.in_(upstream_task_ids),
+                    TI.execution_date == self.execution_date,
+                    TI.state.in_([
+                        State.SUCCESS, State.FAILED, State.UPSTREAM_FAILED]),
+                )
             )
-            count = ti[0][0]
-            if count < len(task._upstream_list):
+            successes, done  = qry[0]
+            if successes < done >= len(task._upstream_list):
```

Comment:
if should actually read  `if successes < done == len(task._upstream_list):` but I didn't know whether it would parse. But yeah, it's basically skipping tasks that have one or many `failed` or `upstream_failed` upstream and they may have other succeeded tasks, but they have to be all finished processing.